### PR TITLE
nixos: enrich systemd.nspawn.* options

### LINKED ISSUE #51076: nixos/nspawn: Do not use  makeUnit to create systemd.nspawn files
https://github.com/NixOS/nixpkgs/blob/5087dcf24759bf16fd3a7b1330b079d72404d7be/nixos/modules/system/boot/systemd-nspawn.nix#L94

It erroreousnly adds options like:

```
       systemd.nspawn.<name>.wantedBy
           Units that want (i.e. depend on) this unit. The standard way to make a unit start by default at boot is to set this option to [
           "multi-user.target" ]. That's despite the fact that the systemd.unit(5) manpage says this option goes in the [Install] section that
           controls the behaviour of systemctl enable. Since such a process is stateful and thus contrary to the design of NixOS, setting this
           option instead causes the equivalent inverse .wants symlink to be present, establishing the same desired relationship in a stateless way.

           Type: list of strings

           Default: [ ]

           Declared by:
               <nixpkgs/nixos/modules/system/boot/systemd-nspawn.nix>

```

Whilst `.nspawn` files are not unit files but _config_ files.

### PR DESCRIPTION
**Pull request by tekeri**
###### Motivation for this change
Add a practical systemd-nspawn configuration options.

Possibly resolves https://github.com/NixOS/nixpkgs/issues/51076 and https://github.com/NixOS/nixpkgs/issues/40367 .

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @c0bw3b @volth @arianvp @Mic92

### CODE REVIEW

**Review by andir — COMMENTED** [review]
It looks good. Could you make sure we have a test for this feature?

**Comment by tekeri** [comment]
I'm new to NixOS tests, so it takes a while.

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 1]
```diff
@@ -42,8 +42,61 @@ let
...
+  mounts = mkOption {
+    type = types.attrsOf (types.submodule {
+      options = {
+        source = mkOption {
+          type = types.str;
+          default = "";
+          description = "mount source path";
+        };
+        mountOption = mkOption {
+          type = types.str;
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix]: Would it make sense to use `types.enum []` here?

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 2]
```diff
@@ -105,29 +171,40 @@ in {
...
       default = {};
       type = with types; attrsOf (submodule instanceOptions);
-      description = "Definition of systemd-nspawn configurations.";
+      description = ''
+        Definition of <citerefentry><refentrytitle>systemd.nspawn</refentrytitle>
+        <manvolnum>5</manvolnum></citerefentry> configurations.
+      '';
+    };
+
+    systemd.customContainers = mkOption {
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix]: IIRC the ExecStart override was needed as we encountered issues with unprivileged users in systemd-nspawn in conjunction with `DynamicUser=true` in services, so this should probably be enabled by default.

Also, I'm not even sure if we should add a toggle for this (but if we keep this, I guess that this should be something like `systemd.disableUnprivilegedNspawn`).

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 1 above)
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix]: I came back. we may specify like 'mount=700,uid=0,gid=0' as a mount options of tmpfs, so enum types would not fit here.

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 2 above)
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix]: > Also, I'm not even sure if we should add a toggle for this 

Exactly, it is not a toggle but just a comment out in my fork. If this breaking change is acceptable, removing this toggle is an option (I'm not sure right direction though).

Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix]: changed to disableUnprivilegedNspawn

**Comment by tekeri** [comment]
@andir test is ready

File: nixos/tests/systemd-nspawn-options.nix [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# Test for NixOS container options.
+
+import ./make-test.nix ({ pkgs, ...} : {
```
Comment (Ma27) [inline nixos/tests/systemd-nspawn-options.nix]: We're actually switching our test suite to python-based tests (this can be used with `./make-test-python.nix`).

File: nixos/tests/systemd-nspawn-options.nix [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      $conf =~ /BindReadOnly=\/nix\/store:\/nix\/store:norbind/ or die;
+      $conf =~ /Bind=\/bind-src:\/bind-dest/ or die;
+      $conf =~ /Bind=\/tmp/ or die;
+      $conf =~ /TemporaryFileSystem=\/tmpfs/ or die;
+      $conf =~ /TemporaryFileSystem=\/var\/tmp:mode=700/ or die;
+      $conf =~ /Inaccessible=\/home/ or die;
+      $conf =~ /Inaccessible=\/root/ or die;
+      $conf =~ /OverlayReadOnly=\/lib:\/usr\/lib64:\/lib64/ or die;
+      $conf =~ /Overlay=\+\/var::\/var/ or die;
+      $conf =~ /ReadOnly=false/ or die;
```
Comment (Ma27) [inline nixos/tests/systemd-nspawn-options.nix]: As we boot into a VM here, those tests usually serve as integration tests, so I'd prefer to see some kind of test here which actually checks if nspawn machines properly start up (for inspiration you might have a look at `nixos/tests/containers-*.nix`).

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 5]
```diff
@@ -87,16 +194,30 @@ let
     let base = {
       text = ''
         [Exec]
+        ${let env = def.environment;
+          in concatMapStrings (n:
+            let s = optionalString (env."${n}" != null)
+              "Environment=${builtins.toJSON "${n}=${env.${n}}"}\n";
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix]: I guess that at least some of those key-value generators could be simplified by using `lib.generators.toKeyValue`.

**Comment by tekeri** [comment]
Thanks, I'll fix them.

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 5 above)
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix]: `toKeyValue` has a bit different behavior... But I could find simplified form in other code locations. Thanks!

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 6]
```diff
@@ -42,8 +42,127 @@ let
...
+      options = {
+        source = mkOption {
+          type = either str path;
+          default = "";
+          description = "Optional mount source path. If omitted, the specified path will be mounted from the host to the same path in the container.";
+        };
+        mountOption = mkOption {
+          type = str;
+          default = "";
+          description = "Optional mount options. See https://www.freedesktop.org/software/systemd/man/systemd-nspawn.html#Mount%20Options for more info.";
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix]: (nitpick) You may want to use `<link xlink:href="url">linktext</link>` here. It will be properly rendered in the man-page and the link will be clickable in the HTML-output of the manual.

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 7]
```diff
@@ -42,8 +42,127 @@ let
     (assertValueOneOf "VirtualEthernet" boolValues)
   ];
 
+  bindMounts = mkOption {
+    type = with types; attrsOf (submodule {
+      options = {
+        source = mkOption {
+          type = either str path;
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix]: The path will be copied to the Nix-Store and the store-path will be mounted when using `types.path` here. We either want to drop it completely or do something like `apply = toString` (`"${./.}"` is a store-path, `"${toString ./.}"` is the (unvalidated) string-representation of the path-value)

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 8]
```diff
@@ -42,8 +42,127 @@ let
...
+  fromVeths = mapAttrsToList (container: host: "${host}:${container}");
+
   instanceOptions = {
-    options = sharedOptions // {
+    options = {
+
+      environment = mkOption {
+        default = {};
+        example = { PATH = "/run/current-system/sw/bin"; };
+        type = with types; attrsOf (nullOr (either str (either path package)));
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix]: Any reason, no `types.int` is allowed here?

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 5 above)
Comment (Ma27) [reply nixos/modules/system/boot/systemd-nspawn.nix]: I see :+1:

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 9]
```diff
@@ -42,8 +42,127 @@ let
...
+                options = {
+                  path = mkOption {
+                    type = either str path;
+                    default = "";
+                    description = "mount destination";
+                  };
+                  mountOption = mkOption {
+                    type = str;
+                    default = "";
+                    description = "mount options. See https://www.freedesktop.org/software/systemd/man/systemd-nspawn.html#Mount%20Options for more info.";
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix]: Ideally, the descriptions should be sentences (with capital letter at the start).

File: nixos/tests/systemd-nspawn-options.nix [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      assert "TemporaryFileSystem=/tmp" in conf
+      assert "TemporaryFileSystem=/var/tmp:mode=777" in conf
+      assert "Inaccessible=/home" in conf
+      assert "Inaccessible=/srv" in conf
+      # assert "OverlayReadOnly=/lib:/usr/lib64:/lib64" in conf
+      # assert "Overlay=+/usr::/usr" in conf
+      assert "ReadOnly=false" in conf
+
+      assert "VirtualEthernetExtra=veth-host:veth-container" in conf
+      assert "Private=true" in conf
```
Comment (Ma27) [inline nixos/tests/systemd-nspawn-options.nix]: IMHO most of those assertions are obsolete when doing more integration-testing

**Review by Ma27 — COMMENTED** [review]
Thanks a lot for your work on this and your patience! I have found a few more notes, but apart from that, the change seems pretty fine. Will test it as soon as I have time to :)

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 11]
```diff
@@ -42,8 +42,127 @@ let
...
+  /*fromBindMounts = mountType: attrs: concatStringsSep "\n"
+    (mapAttrsToList
+      (dest: { source, mountOption, ... }:
+        if mountOption != "" then "${mountType}=${source}:${dest}:${mountOption}"
+        else if source != "" then "${mountType}=${source}:${dest}"
+        else "${mountType}=${dest}")
+      attrs);
+  */
+  fromTmpfsMounts = map (dest:
+    if builtins.isString dest then toString dest
```
[the remark this answers is not in the document]
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix:79]: Exactly.

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 12]
```diff
@@ -42,8 +42,127 @@ let
...
+    else toString dest);
+
+  /*fromBindMounts = mountType: attrs: concatStringsSep "\n"
+    (mapAttrsToList
+      (dest: { source, mountOption, ... }:
+        if mountOption != "" then "${mountType}=${source}:${dest}:${mountOption}"
+        else if source != "" then "${mountType}=${source}:${dest}"
+        else "${mountType}=${dest}")
+      attrs);
+  */
```
[the remark this answers is not in the document]
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix:77]: I'll fix it.

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 6 above)
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix:56]: Thanks, fixed.

BTH, any documentation of the xml?

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 7 above)
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix:49]: Thanks, dropped it for simplicity.

> The path will be copied to the Nix-Store and the store-path will be mounted when using `types.path` here.

A bit surprising for me.

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 8 above)
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix:117]: Nice to add

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 9 above)
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix:138]: Nice catch

File: nixos/tests/systemd-nspawn-options.nix (hunk 10 above)
Comment (tekeri) [reply nixos/tests/systemd-nspawn-options.nix:82]: It seems good to reduce amount

**Comment by flokli** [comment]
@tekeri this needs another rebase.

@Ma27 did you get a chance to test this in the meantime?

**Comment by Ma27** [comment]
@flokli I'm sorry, no. But added it on my todolist for tomorrow :sweat_smile:

**Comment by tekeri** [comment]
pushed, while github is down

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 13]
```diff
@@ -42,8 +42,113 @@ let
...
+          type = str;
+          default = "";
+          description = "Optional mount source path. If omitted, the specified path will be mounted from the host to the same path in the container.";
+        };
+        mountOption = mkOption {
+          type = str;
+          default = "";
+          description = ''
+            Optional mount options. See
+            <link xlink:href="url">https://www.freedesktop.org/software/systemd/man/systemd-nspawn.html#Mount%20Options</link>
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix:58]: shouldn't the `href` attribute contain a URL

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 14]
```diff
@@ -105,29 +226,46 @@ in {
...
+          ) cfg;
+      })
+      { systemd.targets.multi-user.wants = [ "machines.target" ]; }
+      (mkIf config.systemd.disableUnprivilegedNspawn {
+        # Workaround for https://github.com/NixOS/nixpkgs/pull/67232#issuecomment-531315437 and https://github.com/systemd/systemd/issues/13622
+        # Once systemd fixes this upstream, we can re-enable -U
+        systemd.services."systemd-nspawn@".serviceConfig.ExecStart = [
+          ""  # deliberately empty. signals systemd to override the ExecStart
+          # Only difference between upstream is that we do not pass the -U flag
+          "${config.systemd.package}/bin/systemd-nspawn --quiet --keep-unit --boot --link-journal=try-guest --network-veth --settings=override --machine=%i"
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix:267]: @flokli do you know if the systemd bug which made this necessary is fixed? Also, I'd prefer to set `disableUnprivilegedNspawn` to `true` by default since it's been a valid workaround for us.

File: nixos/modules/system/boot/systemd-nspawn.nix [hunk 15]
```diff
@@ -42,8 +42,113 @@ let
...
+                      <link xlink:href="url">https://www.freedesktop.org/software/systemd/man/systemd-nspawn.html#Mount%20Options</link>
+                      for more info.
+                    '';
+                  };
+                };
+              })
+            ]
+          );
+        default = [];
+        example = [ "/tmp" { path = "/var/tmp"; mountOption = ""; } ];
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd-nspawn.nix:131]: Please use `literalExample` for examples that contain a list or an attr-path for better formatting in the manual.

File: nixos/tests/systemd-nspawn-options.nix [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        config = {
+          boot.isContainer = true;
+          networking.useDHCP = false;
+        };
+      })
+    ];
+  };
+in
+
+import ./make-test-python.nix ({ pkgs, ...} : {
```
Comment (Ma27) [inline nixos/tests/systemd-nspawn-options.nix:15]: Do you know why `nscd` inside the container breaks:

```
machine # [   13.949141] systemd-nspawn[786]: [  OK  ] Reached target Basic System.
machine # [   13.953850] systemd-nspawn[786]:          Starting Name Service Cache Daemon...
machine # [   13.960114] systemd-nspawn[786]:          Starting resolvconf update...
machine # [   13.982618] systemd-nspawn[786]: [FAILED] Failed to start Name Service Cache Daemon.
machine # [   13.984544] systemd-nspawn[786]: See 'systemctl status nscd.service' for details.
machine # [   13.988963] systemd-nspawn[786]: [  OK  ] Reached target Host and Network Name Lookups.
machine # [   13.992789] systemd-nspawn[786]: [  OK  ] Reached target User and Group Name Lookups.
machine # [   14.002572] systemd-nspawn[786]:          Starting Login Service...
machine # [   14.054966] systemd-nspawn[786]: [FAILED] Failed to start Login Service.
machine # [   14.057025] systemd-nspawn[786]: See 'systemctl status systemd-logind.service' for details.
machine # [   14.061889] systemd-nspawn[786]: [  OK  ] Stopped Login Service.
machine # [   14.068607] systemd-nspawn[786]:          Starting Login Service...
machine # [   14.094932] systemd-nspawn[786]: [  OK  ] Stopped Name Service Cache Daemon.
machine # [   14.101640] systemd-nspawn[786]:          Starting Name Service Cache Daemon...
machine # [   14.129406] systemd-nspawn[786]: [FAILED] Failed to start Name Service Cache Daemon.
machine # [   14.131745] systemd-nspawn[786]: See 'systemctl status nscd.service' for details.
machine # [   14.154355] systemd-nspawn[786]: [FAILED] Failed to start Login Service.
machine # [   14.156893] systemd-nspawn[786]: See 'systemctl status systemd-logind.service' for details.
machine # [   14.163836] systemd-nspawn[786]: [  OK  ] Stopped Login Service.
machine # [   14.178079] systemd-nspawn[786]:          Starting Login Service...
machine # [   14.243987] systemd-nspawn[786]: [  OK  ] Stopped Name Service Cache Daemon.
machine # [   14.249399] systemd-nspawn[786]:          Starting Name Service Cache Daemon...
machine # [   14.258114] systemd-nspawn[786]: [FAILED] Failed to start Login Service.
machine # [   14.261136] systemd-nspawn[786]: See 'systemctl status systemd-logind.service' for details.
```

File: nixos/tests/systemd-nspawn-options.nix [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+      assert "TemporaryFileSystem=/tmp" in conf
+      assert "TemporaryFileSystem=/var/tmp:mode=777" in conf
+      assert "Inaccessible=/home" in conf
+      assert "VirtualEthernetExtra=veth-host:veth-container" in conf
+      # Fixme: overlayfs won't work on ZFS
+      # assert "OverlayReadOnly=/lib:/usr/lib64:/lib64" in conf
+      # assert "Overlay=+/usr::/usr" in conf
+
+      # Checks filesConfig, networkConfig
+      assert "ReadOnly=false" in conf
```
Comment (Ma27) [inline nixos/tests/systemd-nspawn-options.nix:87]: I don't really think that we need this. As soon as some upstream options change we have to fix this test accordingly. Those kind of assertions don't belong in an integration test IMHO.

File: nixos/tests/systemd-nspawn-options.nix [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+      assert "Private=true" in conf
+
+      machine.succeed(
+          "mkdir -p /var/lib/machines/test-container/{etc,home,srv,tmp,usr,lib} /nix/var/nix/{profiles,gcroots}/per-container/test-container",
+          "chmod 0700 /var/lib/machines /nix/var/nix/{profiles,gcroots}/per-container",
+          "touch /var/lib/machines/test-container/etc/os-release",
+          "machinectl start test-container",
+          "machinectl terminate test-container",
+      )
```
Comment (Ma27) [inline nixos/tests/systemd-nspawn-options.nix:96]: Can you please add a more detailed test to make sure that the machine *actually* works? Some ideas:

* Wait at least for `multi-user.target`.
* Check if networking is fine.
* Check if bind-mounts and restrictions do work.

File: nixos/tests/systemd-nspawn-options.nix (hunk 17 above)
Comment (tekeri) [reply nixos/tests/systemd-nspawn-options.nix:87]: This is actually an unit test to check whether the boolean value is rendered as intended. Also, these lines gives an idea of how the options works.
No problem, however, to remove those assertions.

File: nixos/tests/systemd-nspawn-options.nix (hunk 18 above)
Comment (tekeri) [reply nixos/tests/systemd-nspawn-options.nix:96]: FYI: This changeset is intended to be used with `nixops-container` backend. https://github.com/NixOS/nixops/issues/1226
I'm internally using it and It works over 1 year with networking and bind-mounts.

File: nixos/modules/system/boot/systemd-nspawn.nix (hunk 13 above)
Comment (tekeri) [reply nixos/modules/system/boot/systemd-nspawn.nix:58]: could you tell me how to link to 'Mount options' section in the manual?

**Comment by tekeri** [comment]
I'll fix the conflict and add tests when I have spare time.

**Comment by bqv** [comment]
Interest

**Comment by hmenke** [comment]
I haven't reviewed this PR in detail but I'm strongly in favor of a refactor of `systemd.nspawn`.  Right now it's barely usable.  I had to add some boilerplate which is nowhere to be found in the documentation to make the container boot at all.
```nix
{
  # Boilerplate
  systemd.targets.machines.enable = true;
  systemd.services."systemd-nspawn@archlinux" = {
    enable = true;
    wantedBy = [ "machines.target" ];
  };

  # Actual container
  systemd.nspawn."archlinux" = {
    enable = true; # does this even do anything?
    execConfig = {
      Boot = true;
    };
  };
}
```
But even with these options, the container still refuses to shut down properly on reboot and I'm stuck at the dreaded
```
[    **] A stop job is running for systemd-n…@archlinux.service (7s / 1min 30s)
```
In the journal it leaves
```
Dec 30 19:21:48 nixos systemd[1]: Stopping systemd-nspawn@archlinux.service...
Dec 30 19:21:48 nixos systemd-nspawn[499]: archlinux login: Trying to halt container. Send SIGTERM again to trigger immed>
Dec 30 19:21:48 nixos systemd-nspawn[499]: systemd 245.5-2-arch running in system mode. (+PAM +AUDIT -SELINUX -IMA -APPAR>
Dec 30 19:21:48 nixos systemd-nspawn[499]: Detected virtualization systemd-nspawn.
Dec 30 19:21:48 nixos systemd-nspawn[499]: Detected architecture x86-64.
Dec 30 19:23:18 nixos systemd[1]: systemd-nspawn@archlinux.service: State 'stop-sigterm' timed out. Killing.
```

**Comment by arianvp** [comment]
Im pretty sure the machines.target is enabled by default...

I agree that Systemd-nspawn@.service should have a wantedBy=machines.target
by default though if that isn't the case. To align it with upstream.

**Comment by arianvp** [comment]
Let me rephrase:

I think having `systemd.nspawn.<name>.enable` imply `systemd.services.`systemd-nspawn@<name>'.enable` makes sense. We should add that.

the explicit `wantedBy` is needed because NixOS doesn't handle the upstream `[Install]` sections in unit files

So we should also add that.

That you needed
```
 systemd.targets.machines.enable = true;
```

surprises me though.   `enable = true` is the default for all systemd units.

**Comment by hmenke** [comment]
@arianvp Indeed, I just checked again and `systemd.targets.machines.enable = true;` is not necessary.  However, there is another problem with the systemd integration of NixOS.  By specifying
```nix
{
  systemd.services."systemd-nspawn@archlinux".enable = true;
}
```
NixOS does *not* instantiate the `systemd-nspawn@.service` unit but instead creates a completely new unit named `systemd-nspawn@archlinux.service` which of course doesn't inherit anything from the uninstantiated unit.  So this way of enabling the container is clearly wrong.  To see this better, simply compare and contrast the output of `systemctl cat systemd-nspawn@archlinux.service` with `systemctl cat systemd-nspawn@.service` (or have a look at https://github.com/systemd/systemd/blob/master/units/systemd-nspawn%40.service.in).  This is quite problematic because now essential options for the proper working of the container are missing.

**Comment by tekeri** [comment]
@hmenke I think `systemd-nspawn@$machine.service` is necessary for integration with `machinectl start $machine`.
That is currently implemented as a call of `systemctl start systemd-nspawn@$machine.service` and no way to override the behavior.

**Comment by tekeri** [comment]
it seems out of date.
